src/jit/MemoryInl.h

```diff
@@ -1011,3 +1011,326 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
+    sljit_s32 operation;
+    uint32_t options = MemAddress::AbsoluteAddress;
+
+    switch (instr->opcode()) {
+    case ByteCode::I64AtomicRmwAddOpcode:
+    case ByteCode::I64AtomicRmwSubOpcode:
+    case ByteCode::I64AtomicRmwAndOpcode:
+    case ByteCode::I64AtomicRmwOrOpcode:
+    case ByteCode::I64AtomicRmwXorOpcode:
+    case ByteCode::I64AtomicRmwXchgOpcode: {
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+        emitAtomicRmw64(compiler, instr);
+        return;
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+        operationSize = SLJIT_MOV;
+        size = 8;
+        options |= MemAddress::CheckNaturalAlignment;
+        break;
+    }
+    case ByteCode::I32AtomicRmwAddOpcode:
+    case ByteCode::I32AtomicRmwSubOpcode:
+    case ByteCode::I32AtomicRmwAndOpcode:
+    case ByteCode::I32AtomicRmwOrOpcode:
+    case ByteCode::I32AtomicRmwXorOpcode:
+    case ByteCode::I32AtomicRmwXchgOpcode: {
+        operationSize = SLJIT_MOV32;
+        size = 4;
+        options |= MemAddress::CheckNaturalAlignment;
+        break;
+    }
+    case ByteCode::I64AtomicRmw32AddUOpcode:
+    case ByteCode::I64AtomicRmw32SubUOpcode:
+    case ByteCode::I64AtomicRmw32AndUOpcode:
+    case ByteCode::I64AtomicRmw32OrUOpcode:
+    case ByteCode::I64AtomicRmw32XorUOpcode:
+    case ByteCode::I64AtomicRmw32XchgUOpcode: {
+        operationSize = SLJIT_MOV_U32;
+        size = 4;
+        options |= MemAddress::CheckNaturalAlignment;
+        break;
+    }
+    case ByteCode::I32AtomicRmw8AddUOpcode:
+    case ByteCode::I32AtomicRmw8SubUOpcode:
+    case ByteCode::I32AtomicRmw8AndUOpcode:
+    case ByteCode::I32AtomicRmw8OrUOpcode:
+    case ByteCode::I32AtomicRmw8XorUOpcode:
+    case ByteCode::I32AtomicRmw8XchgUOpcode: {
+        operationSize = SLJIT_MOV32_U8;
+        size = 1;
+        break;
+    }
+    case ByteCode::I64AtomicRmw8OrUOpcode:
+    case ByteCode::I64AtomicRmw8AddUOpcode:
+    case ByteCode::I64AtomicRmw8SubUOpcode:
+    case ByteCode::I64AtomicRmw8AndUOpcode:
+    case ByteCode::I64AtomicRmw8XorUOpcode:
+    case ByteCode::I64AtomicRmw8XchgUOpcode: {
+        operationSize = SLJIT_MOV_U8;
+        size = 1;
+        break;
+    }
+    case ByteCode::I32AtomicRmw16AddUOpcode:
+    case ByteCode::I32AtomicRmw16SubUOpcode:
+    case ByteCode::I32AtomicRmw16AndUOpcode:
+    case ByteCode::I32AtomicRmw16OrUOpcode:
+    case ByteCode::I32AtomicRmw16XorUOpcode:
+    case ByteCode::I32AtomicRmw16XchgUOpcode: {
+        operationSize = SLJIT_MOV32_U16;
+        size = 2;
+        options |= MemAddress::CheckNaturalAlignment;
+        break;
+    }
+    case ByteCode::I64AtomicRmw16AddUOpcode:
+    case ByteCode::I64AtomicRmw16SubUOpcode:
+    case ByteCode::I64AtomicRmw16AndUOpcode:
+    case ByteCode::I64AtomicRmw16OrUOpcode:
+    case ByteCode::I64AtomicRmw16XorUOpcode:
+    case ByteCode::I64AtomicRmw16XchgUOpcode: {
+        operationSize = SLJIT_MOV_U16;
+        size = 2;
+        options |= MemAddress::CheckNaturalAlignment;
```

Comment:
I would set `MemAddress::CheckNaturalAlignment` as the default and set it to 0 in the 8 bit case.